gluster/subvol/cmd/gluster-subvol-provisioner/gluster-subvol-provisioner.go

```diff
@@ -0,0 +1,608 @@
+	// reference of the endpoints name is not sufficient, it can be in an
+	// other namespace.
+	sourceEP, err := p.client.CoreV1().Endpoints(sourceNS).Get(sourcePV.Spec.Glusterfs.EndpointsName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("endpoint %s for source PV %s can not be found: %s", sourcePV.Spec.Glusterfs.EndpointsName, sourcePV.Name, err)
+	}
+
+	ep := sourceEP.DeepCopy()
+	ep.ObjectMeta = metav1.ObjectMeta{
+		Namespace: destNS,
+		Name:      dynamicEpSvcPrefix + destPVCName,
+		Labels: map[string]string{
+			"gluster.org/provisioned-for-pvc": destPVCName,
+		},
+	}
+
+	_, err = p.client.CoreV1().Endpoints(destNS).Create(ep)
+	if err != nil && errors.IsAlreadyExists(err) {
+		glog.V(1).Infof("endpoint %s already exist in namespace %s, that is ok", destPVCName, destNS)
+		err = nil
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to create endpoint %s: %s", ep.ObjectMeta.Name, err)
+	}
+
+	return ep, nil
+}
+
+// Pre-check for the requirements of the PVC-request and the supervol PVC.
+// - certain features of the PVC-request may not be supported
+// - the supervol PVC has some requirements that it needs to fulfil
+func (p *glusterSubvolProvisioner) validateProvisionRequirements(pvcReq, supervolPVC *v1.PersistentVolumeClaim) error {
+	if pvcReq.Spec.Selector != nil {
+		return fmt.Errorf("claim Selector is not supported")
+	}
+
+	if !util.AccessModesContainedInAll(accessModes, pvcReq.Spec.AccessModes) {
+		return fmt.Errorf("invalid AccessModes %v: only AccessModes %v are supported", pvcReq.Spec.AccessModes, accessModes)
+	}
+
+	// check permission mode, the PV should be RWX as it is mounted here,
+	// and we'll create subdirs as new PVCs
+	if !util.AccessModesContains(supervolPVC.Spec.AccessModes, v1.ReadWriteMany) {
+		return fmt.Errorf("this provisioner requires the PVC %s/%s to have ReadWriteMany permissions", supervolPVC.Namespace, supervolPVC.Name)
+	}
+
+	return nil
+}
+
+// Check if the PVC has a CloneRequest annotation and try to clone if it has.
+// This returns the source namespace/PVC that got cloned, an error if something fails.
+func (p *glusterSubvolProvisioner) tryClone(pvc *v1.PersistentVolumeClaim, mountpoint, destDir string) (string, error) {
+	// sourcePVCRef points to the PVC that should get cloned
+	sourcePVCRef, ok := pvc.Annotations[CloneRequestAnn]
+	if !ok || sourcePVCRef == "" {
+		// no CloneRequest, no need to try and clone
+		return "", nil
+	}
+
+	glog.Infof("got requested to clone %s for PVC %s/%s", sourcePVCRef, pvc.Namespace, pvc.Name)
+
+	// sourcePVCRef is like (namespace/)?pvc
+	var sourceNS, sourcePVCName string
+
+	parts := strings.Split(sourcePVCRef, "/")
+	if len(parts) == 1 {
+		sourceNS = pvc.Namespace
+		sourcePVCName = parts[0]
+	} else if len(parts) == 2 {
+		sourceNS = parts[0]
+		sourcePVCName = parts[1]
+	} else {
+		return "", fmt.Errorf("failed to parse namespace/pvc from %s", sourcePVCRef)
+	}
+
+	// TODO: check if the size of the PVC >= source
```

Comment:
@nixpanic can you address this TODO ?

Reply:
Oh, this TODO is actually done. See https://github.com/kubernetes-incubator/external-storage/pull/1013/files#diff-066351a7c1b20ad6a21c9ad39e1257e4R375 line 375. I'll drop it for the next updated version of this PR.